commons/zenoh-shm/src/api/provider/shared_memory_provider.rs

```diff
@@ -0,0 +1,910 @@
+}
+impl<'a, IDSource, Backend> AllocLayoutBuilder<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    /// Set size for layout
+    #[zenoh_macros::unstable_doc]
+    pub fn size(self, size: usize) -> AllocLayoutSizedBuilder<'a, IDSource, Backend> {
+        AllocLayoutSizedBuilder {
+            provider: self.provider,
+            size,
+        }
+    }
+}
+
+pub struct AllocLayoutSizedBuilder<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    provider: &'a SharedMemoryProvider<IDSource, Backend>,
+    size: usize,
+}
+impl<'a, IDSource, Backend> AllocLayoutSizedBuilder<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    /// Set alignment for layout
+    #[zenoh_macros::unstable_doc]
+    pub fn alignment(
+        self,
+        alignment: AllocAlignment,
+    ) -> AllocLayoutAlignedBuilder<'a, IDSource, Backend> {
+        AllocLayoutAlignedBuilder {
+            provider: self.provider,
+            size: self.size,
+            alignment,
+        }
+    }
+
+    /// try to build an allocation layout
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> ZResult<AllocLayout<'a, IDSource, Backend>> {
+        AllocLayout::new(self.size, AllocAlignment::default(), self.provider)
+    }
+}
+
+pub struct AllocLayoutAlignedBuilder<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    provider: &'a SharedMemoryProvider<IDSource, Backend>,
+    size: usize,
+    alignment: AllocAlignment,
+}
+impl<'a, IDSource, Backend> AllocLayoutAlignedBuilder<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    /// Try to build layout with specified args
+    #[zenoh_macros::unstable_doc]
+    pub fn res(self) -> ZResult<AllocLayout<'a, IDSource, Backend>> {
+        AllocLayout::new(self.size, self.alignment, self.provider)
+    }
+}
+
+/// A layout for allocations.
+/// This is a pre-calculated layout suitable for making series of similar allocations
+/// adopted for particular SharedMemoryProvider
+#[zenoh_macros::unstable_doc]
+#[derive(Debug)]
+pub struct AllocLayout<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    size: usize,
+    provider_layout: MemoryLayout,
+    provider: &'a SharedMemoryProvider<IDSource, Backend>,
+}
+
+impl<'a, IDSource, Backend> AllocLayout<'a, IDSource, Backend>
+where
+    IDSource: ProtocolIDSource,
+    Backend: SharedMemoryProviderBackend,
+{
+    /// Allocate the new buffer with this layout
+    #[zenoh_macros::unstable_doc]
+    pub fn alloc(&'a self) -> AllocBuilder<'a, IDSource, Backend> {
+        AllocBuilder {
+            layout: self,
+            _phantom: PhantomData,
+        }
+    }
+
```

Comment:
```suggestion

#[zenoh_macros::unstable_doc]
```

Reply:
`new()` is private, so it is not documented

Reply:
Right, irrelevant comment.